libs/test-utils/src/expect_print.ts

```diff
@@ -0,0 +1,144 @@
+    throw new ExpectPrintError(
+      'You have not made any assertions against the last print before printing another within a test.'
+    );
+  }
+  lastPrintedElement = element;
+  lastPrintOptions = printOptions;
+  return Promise.resolve();
+}
+
+export function fakePrintElementWhenReady(
+  elementWithCallback: ElementWithCallback,
+  printOptions: PrintOptions
+): Promise<void> {
+  if (lastPrintedElement || lastPrintedElementWithCallback) {
+    throw new ExpectPrintError(
+      'You have not made any assertions against the last print before printing another within a test.'
+    );
+  }
+  lastPrintedElementWithCallback = elementWithCallback;
+  lastPrintOptions = printOptions;
+  return Promise.resolve();
+}
+
+/**
+ * Renders the element passed to fakePrintElement and allows assertions
+ * against its content.
+ *
+ * @param inspect method which runs queries against the render result
+ */
+function inspectPrintedElement(inspect: InspectPrintFunction): void {
+  assert(lastPrintedElement);
+  const renderResult = render(lastPrintedElement);
+  inspect(renderResult, lastPrintOptions);
+  renderResult.unmount();
+}
+
+/**
+ * Renders the element with callback passed to fakePrintElementWhenReady
+ * and allows assertions against its content.
+ *
+ * @param inspect method which runs queries against the render result
+ */
+async function inspectPrintedWhenReadyElement(
+  inspect: InspectPrintFunction
+): Promise<void> {
+  let renderResult: Optional<RenderResult>;
+  const renderedPromise = new Promise<void>((resolve) => {
+    assert(lastPrintedElementWithCallback);
+    renderResult = render(lastPrintedElementWithCallback(resolve));
+  });
+  // We have to advance promises here to allow effects to run before
+  // running assertions against the rendered element.
+  await advancePromises();
+  await renderedPromise;
+  assert(renderResult);
+  inspect(renderResult, lastPrintOptions);
+  renderResult.unmount();
+}
+
+/**
+ * Asserts that there are no elements which were mock printed and have not been
+ * tested. Should run after tests to confirm there were no unexpected prints.
+ */
+export function expectAllPrintsAsserted(): void {
+  if (lastPrintedElement || lastPrintedElementWithCallback) {
+    throw new ExpectPrintError(
+      `This test ended with prints that were not asserted against.`
+    );
+  }
+}
+
+export function resetExpectPrint(): void {
+  lastPrintedElement = undefined;
+  lastPrintedElementWithCallback = undefined;
+  lastPrintOptions = undefined;
+}
+
+/**
+ * Asserts that a print took place via the mocks of printElement or
+ * printElementWhenReady. Renders the printed element and allows assertions
+ * against its content. When using this utility, you should assert against
+ * all prints.
+ *
+ * @param inspect method which runs queries against the render result
+ */
+export async function expectPrint(
+  inspect?: InspectPrintFunction
+): Promise<void> {
+  await advancePromises();
```

Comment:
This `advance` is not strictly necessary, but it's nice for the consumer not to have to do it themselves.

The issue with waiting on the variable to be populated is that it may not be, and we could hang on that promise. We could set a timeout but them we're essentially doing the same thing as just an empty `waitFor` plus more.

I understand the reservation (this pattern has always seemed weird to me) and I would be interested in reading anything RTL has written about it, but for now I'd prefer to keep it here.

Reply:
`waitFor` has a built-in timeout, so I think it should be safe

Reply:
I tried it and I don't want to use `waitFor` to actually wait for something because that would require using an `expect` under the hood. I think it's confusing for a test utility to make its own assertions under the hood. It throws off the expected assertion count, as I discovered when playing with this. And then the error throwing is done by `jest` and is not in our control.

And, my understanding is that waiting on the `lastPrinted...` is mostly semantic, because in either case the event loop will advance. I don't understand enough about the order of execution to see how the behavior could differ. Except that `waitFor` will first try to run the callback without advancing the event loop, which we could also do by checking for the truthiness of the elements first, but I think that is confusing because then we're only advancing the event loop _sometimes_.

I plan to leave this as-is unless you feel strongly otherwise.

Reply:
Hmm, I don't think I agree that `expectPrint` shouldn't make assertions. It's got `expect` in the name! And the next thing it does after this is make an assertion... even though it's using `ExpectPrintError` instead of a jest assertion. In fact, I don't think `waitFor` requires you to use a jest assertion, so if that's a concern, you can just use the `ExpectPrintError` instead.

I'm not sure what you mean about the "expected assertion count" - what is that?

Reply:
When testing assertions made in `catch` blocks, you will need to expect that an assertion was made (in addition the assertion itself) to know that it actually ran. E.g. here: https://github.com/votingworks/vxsuite/blob/00c111f88ccd04b78f5ecc0e16fb366d18039871/libs/test-utils/src/expect_print.test.ts#L124. `expectPrint` should make assertions, but shouldn't use `expect`. That is what I meant. I didn't know you could exit `waitFor` by throwing any error. In that case, we can just move the error throwing within a `waitFor`. That works. Pushed that change.

Reply:
Hm, I don't see a `catch` block in the code you linked, but I did find one elsewhere in that file. I've never actually used that approach before for testing async errors. TIL. In the past, I've used the `.rejects` matcher